### ISSUE TITLE
Results does not preserve the order of tasks

### ISSUE THREAD

**Issue by bruceCzK** [supercharge/promise-pool#26]
In some minor cases, this'll cause some serious problem. I'm suggesting instead of push result to `results` array, use index to set the result and a `null` for the fail task. Here's the proposed changes to `promise-pool-executor.ts`.
```
  async process (): Promise<ReturnValue<T, R>> {
    for (let i = 0; i < this.items.length; i++) {
      const item = this.items[i]
      if (this.hasReachedConcurrencyLimit()) {
        await this.processingSlot()
      }

      this.startProcessing(item, i) // pass index to startProcessing
    }

    return this.drained()
  }

  startProcessing (item: T, index: Number): void {
    const task = this.createTaskFor(item)
      .then(result => {
        this.results[index] = result // set by index
        this.tasks.splice(this.tasks.indexOf(task), 1)
      })
      .catch(error => {
        this.results[index] = null // set result as null if failed
        this.tasks.splice(this.tasks.indexOf(task), 1)

        if (this.errorHandler) {
          return this.errorHandler(error, item)
        }

        this.errors.push(
          PromisePoolError.createFrom(error, item)
        )
      })

    this.tasks.push(task)
  }
```

**Comment by marcuspoehls** [comment]
@bruceCzK Hey Bruce, thank you for suggesting the ordering improvement. I’m open to adding this feature. Do you want to submit a PR?

**Comment by bruceCzK** [comment]
Oh, I'm not quite familiar with typescript, sorry, I think I'll just pass.

**Comment by marcuspoehls** [comment]
@bruceCzK Bruce, did you run into the ordering issue? If yes, can you share more details about the requirements to keep the index?

**Comment by bruceCzK** [comment]
I have a list of URLs to download as HTML (about 100) and a list of their filenames (article titles), the two lists are only bounded with index. I was using this library, but found out that the order did not preseve, the filenames did not match its content.
I have look into the source code yesterday, it seems it requires to change several test cases, which might be too much change to the current function, made me wondering if this feather is neccessary. What do you think?

**Comment by marcuspoehls** [comment]
@bruceCzK I was thinking about your issue over the last day. I guess you can simply work around this issue by using a single list of objects. And the objects then contain the URL and file name:

```js
const items = [
  {
    url: 'your-url',
    file: '/path/to/the/file/of/your-url.txt'
  }
]
```

This way you don’t need to sync the results with any of your lists of URLs and files. It can be part of the result.

Is that an option for you?

**Comment by bruceCzK** [comment]
Yes. This is the workaround I'm using now. Thanks for the advice. I guess we can both agree this is not a neccesary feather for the time being. Thanks again for your quick reply.

**Comment by tylercollier** [comment]
Great library! I like the API, except for this behavior. I find it confusing. It cost me an hour yesterday. To me it violates the principle of least surprise. I expected the behavior to be like `Promise.all()`, putting the results in the same indexes as the inputs. As it stands, I'm not sure how useful the `results` are from this example from the README:

```js
const { results, errors } = await PromisePool...
```

To link up the inputs and results, best case, there is a unique ID on each input and result, and you have to run `results.find()` to get the corresponding result. But what if there is no unique ID? You can't tell which result corresponds to which input.

Would it be hard or otherwise problematic to put the results in the corresponding index? It seems that @bruceCzK already suggested an approach.

Otherwise, I think it should be documented. I think the error handling caveat is called out well in the README, and that same style of call-out should be used to mention the indeterminate order of results.

**Comment by marcuspoehls** [comment]
@tylercollier Thanks for sharing your thoughts. I'm still open to a merge request if you want to prepare one. I was thinking about the ordering issue and we might easily get it into the package: when adding a task result to the results array, I'm currently pushing the result to the end of the list. What we should do instead is splicing the result at the array position of the related item.

**Comment by tylercollier** [comment]
I was going to take a crack at this. But it says you self assigned it?

**Comment by tylercollier** [comment]
Just a head's up that I submitted a PR for this

**Comment by marcuspoehls** [comment]
@tylercollier Hey Tyler, thank you for the pull request and following up. I’ll look through the PR today

**Comment by popadi** [comment]
What is the status of this? I am currently dealing with the same thing.

**Comment by marcuspoehls** [comment]
This feature is available in version `2.4.0`. Docs are here -> https://superchargejs.com/docs/3.x/promise-pool#correspond-source-items-and-their-results
